engine/execution/ingestion/fetcher/fetcher.go

```diff
@@ -0,0 +1,78 @@
+package fetcher
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+var onlyOnflowRegex = regexp.MustCompile(`.*\.onflow\.org:3569$`)
+
+type CollectionFetcher struct {
+	request module.Requester // used to request collections
+	state   protocol.State
+	// This is included to temporarily work around an issue observed on a small number of ENs.
+	// It works around an issue where some collection nodes are not configured with enough
+	// this works around an issue where some collection nodes are not configured with enough
+	// file descriptors causing connection failures.
+	onflowOnlyLNs bool
+	log           zerolog.Logger
+}
+
+func NewCollectionFetcher(request module.Requester, state protocol.State, onflowOnlyLNs bool, log zerolog.Logger) *CollectionFetcher {
+	return &CollectionFetcher{
+		request:       request,
+		state:         state,
+		onflowOnlyLNs: onflowOnlyLNs,
+		log:           log.With().Str("component", "ingestion_engine_collecition_fetcher").Logger(),
+	}
+}
+
+func (e *CollectionFetcher) FetchCollection(blockID flow.Identifier, height uint64, guarantee *flow.CollectionGuarantee) error {
+	guarantors, err := protocol.FindGuarantors(e.state, guarantee)
+	if err != nil {
+		// execution node executes certified blocks, which means there is a quorum of consensus nodes who
+		// have validated the block payload. And that validation includes checking the guarantors are correct.
+		// Based on that assumption, failing to find guarantors for guarantees contained in an incorporated block
+		// should be treated as fatal error
+		e.log.Fatal().Err(err).Msgf("failed to find guarantors for guarantee %v at block %v, height %v",
+			guarantee.ID(),
+			blockID,
+			height,
+		)
+		return fmt.Errorf("could not find guarantors: %w", err)
```

Comment:
e.log.Fatal calls `os.Exit(1)`. Is that intended here? We might want to use signaller Context here in the future.  Also the return line is never reached.